File: src/devtools/discovery.py
from __future__ import annotations

import json
from dataclasses import asdict
from pathlib import Path

from devtools.models import AppConfig, PROJECT_MARKERS, Project


def detect_project_type(project_path: Path, enabled_types: list[str]) -> tuple[str, str] | None:
    for project_type in enabled_types:
        marker = PROJECT_MARKERS.get(project_type)
        if marker and (project_path / marker).exists():
            return project_type, marker
    return None


def detect_project_types_recursive(project_path: Path, enabled_types: list[str]) -> list[str]:
    found: list[str] = []
    remaining = set(enabled_types)
    for current_root, dirnames, filenames in _walk_project_tree(project_path):
        filename_set = set(filenames)
        for project_type in enabled_types:
            if project_type not in remaining:
                continue
            marker = PROJECT_MARKERS.get(project_type)
            if marker and marker in filename_set:
                found.append(project_type)
                remaining.remove(project_type)
        if not remaining:
            break
    return found


def discover_projects(config: AppConfig, roots: list[Path] | None = None, refresh: bool = False) -> list[Project]:
    if refresh or not config.discovery.cache_file.exists():
        projects = rebuild_project_cache(config, roots=roots)
    else:
        projects = load_project_cache(config)
    return sorted(projects, key=lambda item: (item.name.lower(), item.path.as_posix()))


def rebuild_project_cache(config: AppConfig, roots: list[Path] | None = None) -> list[Project]:
    scan_roots = roots or config.discovery.roots
    projects: dict[str, Project] = {}
    for root in scan_roots:
        if not root.exists():
            continue
        for child in sorted(root.iterdir(), key=lambda item: item.name.lower()):
            if not child.is_dir():
                continue
            if _should_skip_path(child, config):
                continue
            if not _is_included(child, config):
                continue
            detected_types = detect_project_types_recursive(child, config.discovery.project_types)
            if not detected_types:
                continue
            project_type = detected_types[0]
            marker = PROJECT_MARKERS[project_type]
            project = Project(
                name=child.name,
                path=child.resolve(),
                project_type=project_type,
                marker=marker,
                project_types=detected_types,
            )
            projects[project.identity] = project

    _write_project_cache(config, list(projects.values()))
    return list(projects.values())


def load_project_cache(config: AppConfig) -> list[Project]:
    cache_file = config.discovery.cache_file
    with cache_file.open("r", encoding="utf-8") as handle:
        raw = json.load(handle)

    projects: list[Project] = []
    for item in raw.get("projects", []):
        path = Path(item["path"]).expanduser().resolve()
        detected = detect_project_type(path, config.discovery.project_types)
        if detected is None:
            continue
        project_types = detect_project_types_recursive(path, config.discovery.project_types)
        projects.append(
            Project(
                name=str(item["name"]),
                path=path,
                project_type=str(item["project_type"]),
                marker=str(item["marker"]),
                project_types=[str(value) for value in item.get("project_types", project_types or [str(item["project_type"])])],
            )
        )
    return projects


def discover_explicit_projects(config: AppConfig, paths: list[Path]) -> list[Project]:
    projects: list[Project] = []
    for path in paths:
        resolved = path.expanduser().resolve()
        detected = detect_project_type(resolved, config.discovery.project_types)
        if detected is None:
            continue
        project_type, marker = detected
        projects.append(
            Project(
                name=resolved.name,
                path=resolved,
                project_type=project_type,
                marker=marker,
                project_types=detect_project_types_recursive(resolved, config.discovery.project_types),
            )
        )
    return sorted(projects, key=lambda item: (item.name.lower(), item.path.as_posix()))


def filter_projects(
    projects: list[Project],
    project_type: str | None = None,
    name_filter: str | None = None,
) -> list[Project]:
    filtered = projects
    if project_type:
        filtered = [project for project in filtered if project.project_type == project_type]
    if name_filter:
        lowered = name_filter.lower()
        filtered = [
            project
            for project in filtered
            if lowered in project.name.lower() or lowered in project.path.as_posix().lower()
        ]
    return filtered


def _write_project_cache(config: AppConfig, projects: list[Project]) -> None:
    cache_file = config.discovery.cache_file
    cache_file.parent.mkdir(parents=True, exist_ok=True)
    payload = {
        "roots": [root.as_posix() for root in config.discovery.roots],
        "projects": [
            {
                **asdict(project),
                "path": project.path.as_posix(),
            }
            for project in sorted(projects, key=lambda item: (item.name.lower(), item.path.as_posix()))
        ],
    }
    cache_file.write_text(json.dumps(payload, indent=2), encoding="utf-8")


def _is_included(path: Path, config: AppConfig) -> bool:
    include_patterns = config.discovery.include_patterns
    if not include_patterns:
        return True
    return any(path.match(pattern) for pattern in include_patterns)


def _should_skip_path(path: Path, config: AppConfig) -> bool:
    path_text = path.as_posix()
    return any(
        path.match(pattern) or path_text.endswith(pattern.replace("**/", ""))
        for pattern in config.discovery.exclude_patterns
    )


def _walk_project_tree(project_root: Path):
    import os

    for current_root, dirnames, filenames in os.walk(project_root):
        current_path = Path(current_root)
        dirnames[:] = [
            dirname
            for dirname in dirnames
            if not _is_runtime_skip_dir(current_path / dirname)
        ]
        yield current_root, dirnames, filenames


def _is_runtime_skip_dir(path: Path) -> bool:
    return path.name in {"node_modules", ".git", "dist", "target", ".angular", ".cache", ".next"}

File: src/discovery.ts
import fs from "node:fs";
import path from "node:path";

import type { AppConfig, Project, ProjectType } from "./models.ts";
import { PROJECT_MARKERS } from "./models.ts";
import { pathIdentity, toPosixPath } from "./path-utils.ts";

interface CacheFileShape {
  roots: string[];
  projects: Omit<Project, "identity">[];
}

const RUNTIME_SKIP_DIRS = new Set(["node_modules", ".git", "dist", "target", ".angular", ".cache", ".next"]);

export function detectProjectType(projectPath: string, enabledTypes: ProjectType[]): [ProjectType, string] | null {
  for (const projectType of enabledTypes) {
    const marker = PROJECT_MARKERS[projectType];
    if (fs.existsSync(path.join(projectPath, marker))) {
      return [projectType, marker];
    }
  }
  return null;
}

export function detectProjectTypesRecursive(projectPath: string, enabledTypes: ProjectType[]): ProjectType[] {
  const found: ProjectType[] = [];
  const remaining = new Set(enabledTypes);

  walkProjectTree(projectPath, (_dir, filenames) => {
    const fileSet = new Set(filenames);
    for (const projectType of enabledTypes) {
      if (!remaining.has(projectType)) {
        continue;
      }
      const marker = PROJECT_MARKERS[projectType];
      if (fileSet.has(marker)) {
        found.push(projectType);
        remaining.delete(projectType);
      }
    }
    return remaining.size > 0;
  });

  return found;
}

export function discoverProjects(config: AppConfig, roots?: string[], refresh = false): Project[] {
  if (refresh || !fs.existsSync(config.discovery.cacheFile)) {
    return sortProjects(rebuildProjectCache(config, roots));
  }
  return sortProjects(loadProjectCache(config));
}

export function rebuildProjectCache(config: AppConfig, roots?: string[]): Project[] {
  const scanRoots = roots ?? config.discovery.roots;
  const projects = new Map<string, Project>();
  for (const root of scanRoots) {
    if (!fs.existsSync(root)) {
      continue;
    }
    const children = fs.readdirSync(root, { withFileTypes: true }).sort((a, b) => a.name.localeCompare(b.name));
    for (const child of children) {
      if (!child.isDirectory()) {
        continue;
      }
      const childPath = path.resolve(root, child.name);
      if (shouldSkipPath(childPath, config) || !isIncluded(childPath, config)) {
        continue;
      }
      const detectedTypes = detectProjectTypesRecursive(childPath, config.discovery.projectTypes);
      if (detectedTypes.length === 0) {
        continue;
      }
      const projectType = detectedTypes[0];
      const marker = PROJECT_MARKERS[projectType];
      const project: Project = {
        name: child.name,
        path: childPath,
        projectType,
        marker,
        projectTypes: detectedTypes,
        identity: pathIdentity(child.name, childPath),
      };
      projects.set(project.identity, project);
    }
  }
  writeProjectCache(config, [...projects.values()]);
  return [...projects.values()];
}

export function loadProjectCache(config: AppConfig): Project[] {
  const raw = JSON.parse(fs.readFileSync(config.discovery.cacheFile, "utf8")) as CacheFileShape;
  const projects: Project[] = [];
  for (const item of raw.projects ?? []) {
    const resolved = path.resolve(item.path);
    const detected = detectProjectType(resolved, config.discovery.projectTypes);
    if (!detected) {
      continue;
    }
    projects.push({
      ...item,
      path: resolved,
      projectTypes: item.projectTypes?.length ? item.projectTypes : detectProjectTypesRecursive(resolved, config.discovery.projectTypes),
      identity: pathIdentity(item.name, resolved),
    });
  }
  return projects;
}

export function discoverExplicitProjects(config: AppConfig, pathsToInspect: string[]): Project[] {
  const projects: Project[] = [];
  for (const item of pathsToInspect) {
    const resolved = path.resolve(item);
    const detected = detectProjectType(resolved, config.discovery.projectTypes);
    if (!detected) {
      continue;
    }
    const [projectType, marker] = detected;
    projects.push({
      name: path.basename(resolved),
      path: resolved,
      projectType,
      marker,
      projectTypes: detectProjectTypesRecursive(resolved, config.discovery.projectTypes),
      identity: pathIdentity(path.basename(resolved), resolved),
    });
  }
  return sortProjects(projects);
}

export function filterProjects(projects: Project[], projectType?: string, nameFilter?: string): Project[] {
  let filtered = projects;
  if (projectType) {
    filtered = filtered.filter((project) => project.projectTypes.includes(projectType as ProjectType));
  }
  if (nameFilter) {
    const lowered = nameFilter.toLowerCase();
    filtered = filtered.filter((project) => {
      const projectPath = toPosixPath(project.path).toLowerCase();
      return project.name.toLowerCase().includes(lowered) || projectPath.includes(lowered);
    });
  }
  return filtered;
}

function writeProjectCache(config: AppConfig, projects: Project[]): void {
  fs.mkdirSync(path.dirname(config.discovery.cacheFile), { recursive: true });
  const payload: CacheFileShape = {
    roots: [...config.discovery.roots],
    projects: sortProjects(projects).map(({ identity: _identity, ...project }) => project),
  };
  fs.writeFileSync(config.discovery.cacheFile, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
}

function sortProjects(projects: Project[]): Project[] {
  return [...projects].sort((left, right) => {
    const byName = left.name.localeCompare(right.name);
    if (byName !== 0) {
      return byName;
    }
    return toPosixPath(left.path).localeCompare(toPosixPath(right.path));
  });
}

function isIncluded(value: string, config: AppConfig): boolean {
  if (config.discovery.includePatterns.length === 0) {
    return true;
  }
  const posix = toPosixPath(value);
  return config.discovery.includePatterns.some((pattern) => posix.includes(pattern.replace("**/", "")));
}

function shouldSkipPath(value: string, config: AppConfig): boolean {
  const posix = toPosixPath(value);
  return config.discovery.excludePatterns.some((pattern) => posix.includes(pattern.replace("**/", "")));
}

function walkProjectTree(
  root: string,
  visitor: (currentDir: string, filenames: string[]) => boolean,
): void {
  const stack = [root];
  while (stack.length > 0) {
    const currentDir = stack.pop()!;
    const entries = fs.readdirSync(currentDir, { withFileTypes: true });
    const directories: string[] = [];
    const filenames: string[] = [];
    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (!RUNTIME_SKIP_DIRS.has(entry.name)) {
          directories.push(path.join(currentDir, entry.name));
        }
      } else if (entry.isFile()) {
        filenames.push(entry.name);
      }
    }
    const shouldContinue = visitor(currentDir, filenames);
    if (!shouldContinue) {
      return;
    }
    directories.sort((a, b) => a.localeCompare(b)).reverse().forEach((directory) => stack.push(directory));
  }
}
